### PR TITLE
2269-2

### PR DESCRIPTION
**Pull request by Evenstein**
## Номер
2269

## Номер задания
2

## Ссылка на видео с демо
https://youtu.be/2e1AZfzsmOA

## Комментарии
Сделан 1 уровень. UPD: Добавлен 2 уровень.

### CODE REVIEW

File: 2269/2/.gitkeep [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (allomov) [inline 2269/2/.gitkeep]: гэты файл можна выдаліць

File: 2269/2/bad word dict [hunk 2]
```diff
@@ -0,0 +1 @@
+х*й п**д еба* бл* на**й ебал х*есо еб* уе*ан п*зд хуй блять блядь пизд еба нахуй
```
Comment (allomov) [inline 2269/2/bad word dict]: it's better to name files without spaces (use  _ or - instead).

File: 2269/2/no words [hunk 3]
```diff
@@ -0,0 +1,4 @@
+в на о между за под по с без до для у ради к про через сквозь над перед при не ни нет никак пусть
+пускай давайте да бы б бывало вот вон давай а и именно также тоже как если где когда куда откуда
+почему лишь едва что чтобы я ты мы меня мой твой тебя он она оно они но это этот тут все там кто
+так ведь же то ну – его себя нету от из
```
Comment (allomov) [inline 2269/2/no words]: you can use more suitable name for the file. also you can move all config to `.yml` configuration file

File: 2269/2/raper.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# :reek:Attribute
+# :reek:TooManyInstanceVariables
+class Raper
+  def initialize(a_name)
+    @name = a_name
+    @words_round = 0
+    @words_battle = 0
+    @bad_words = 0
+    @battles = 0
+    @file_name = []
```
Comment (allomov) [inline 2269/2/raper.rb]: use memoization for instance variables.

File: 2269/2/rapers_list.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# :reek:DuplicateMethodCall
+# :reek:UtilityFunction
+# :reek:TooManyStatements
+# rubocop:disable Metrics/AbcSize
+# rubocop:disable Metrics/MethodLength
+def get_rapers(destination)
+  hash = {}
+  Dir.entries(destination).each do |file|
+    next if ['.', '..'].include?(file)
```
Comment (allomov) [inline 2269/2/rapers_list.rb]: you can use Dir method that does not include directories.

File: 2269/2/rapers_list.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# :reek:DuplicateMethodCall
+# :reek:UtilityFunction
+# :reek:TooManyStatements
+# rubocop:disable Metrics/AbcSize
+# rubocop:disable Metrics/MethodLength
+def get_rapers(destination)
+  hash = {}
+  Dir.entries(destination).each do |file|
```
Comment (allomov) [inline 2269/2/rapers_list.rb]: that's file_name, not a file object.
ps.sorry being nerdy here, but that's part of the proficiency.

File: 2269/2/rapers_list.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# :reek:DuplicateMethodCall
```
Comment (allomov) [inline 2269/2/rapers_list.rb]: reduce the amount of disabled checks

File: 2269/2/top_bad.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
+# This method returns top Rapers uses bad words
```
Comment (allomov) [inline 2269/2/top_bad.rb]: reduce the amount of disabled checks

File: 2269/2/top_bad.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+# :reek:DuplicateMethodCall
+# :reek:NestedIterators
+# :reek:FeatureEnvy
+# :reek:TooManyStatements
+# rubocop:disable Metrics/AbcSize
+# rubocop:disable Metrics/MethodLength
+def get_top_bad(top, destination)
```
Comment (allomov) [inline 2269/2/top_bad.rb]: unclear method name.

the method is too long.

собака должна объяснить все претензии к этому методу

File: 2269/2/top_bad.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+      hash = word_counter(destination + '/' + file)
+      value.add_words_count hash[:total]
+      value.add_bad_words hash[:bad]
+      value.words_round = hash[:round]
+    end
+  end
+
+  delimiter = 1
+  puts "\n"
+  rapers.sort_by { |_k, val| [-val.bad_words] }.each do |_key, value|
```
Comment (allomov) [inline 2269/2/top_bad.rb]: unclear

File: 2269/2/top_word_counter.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
+# :reek:DuplicateMethodCall
```
Comment (allomov) [inline 2269/2/top_word_counter.rb]: reduce the amount of disabled checks.

File: 2269/2/bad word dict (hunk 2 above)
Comment (Evenstein) [reply 2269/2/bad word dict]: fixed

File: 2269/2/.gitkeep (hunk 1 above)
Comment (Evenstein) [reply 2269/2/.gitkeep]: файл выдалены)

File: 2269/2/no words (hunk 3 above)
Comment (Evenstein) [reply 2269/2/no words]: created config.yml and all data from bad_word_dict and no_words has been moved to this file

File: 2269/2/raper.rb (hunk 4 above)
Comment (Evenstein) [reply 2269/2/raper.rb]: initialiazation of instance variables has been changed to memoization

File: 2269/2/rapers_list.rb (hunk 5 above)
Comment (Evenstein) [reply 2269/2/rapers_list.rb]: this issue fixed

File: 2269/2/rapers_list.rb (hunk 7 above)
Comment (Evenstein) [reply 2269/2/rapers_list.rb]: reduced

File: 2269/2/top_bad.rb (hunk 10 above)
Comment (Evenstein) [reply 2269/2/top_bad.rb]: removed

File: 2269/2/rapers_list.rb (hunk 6 above)
Comment (Evenstein) [reply 2269/2/rapers_list.rb]: changed a method name

File: 2269/2/top_bad.rb (hunk 8 above)
Comment (Evenstein) [reply 2269/2/top_bad.rb]: fixed completely

File: 2269/2/top_bad.rb (hunk 9 above)
Comment (Evenstein) [reply 2269/2/top_bad.rb]: fixed completely

File: 2269/2/total_words.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    helper.add_word_count
+    helper.add_word_count_round
+
+    helper.bad_words_array.each do |bw|
+      helper.add_bad_word_count if word.downcase.include? bw.downcase
+    end
+  end
+end
+
+def round?(helper, line)
```
Comment (allomov) [inline 2269/2/total_words.rb]: methods with `?` should not change the state

File: 2269/2/raper.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
+# Class represents Info about Raper
+class BattleInfo
+  def initialize
+    @words_round ||= 0
+    @words_battle ||= 0
+    @bad_words ||= 0
+    @file_name ||= []
```
Comment (allomov) [inline 2269/2/raper.rb]: зачем тут `||=`?

File: 2269/2/raper.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+  def file_source_info(file)
+    @file_name << file
+  end
+end
+
+# Class that represents each Rap Singer.
+class Raper
+  def initialize(name)
+    @battles ||= 0
```
Comment (allomov) [inline 2269/2/raper.rb]: battles это список? список чего?

File: 2269/2/raper.rb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  def initialize
+    @words_round ||= 0
+    @words_battle ||= 0
+    @bad_words ||= 0
+    @file_name ||= []
+  end
+
+  attr_reader :words_battle, :bad_words, :words_round, :file_name
+
+  def set_result(total, bad, round)
```
Comment (allomov) [inline 2269/2/raper.rb]: `set` implies you assign, not mutate state.

File: 2269/2/raper.rb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    @words_round = round
+  end
+
+  def file_source_info(file)
+    @file_name << file
+  end
+end
+
+# Class that represents each Rap Singer.
+class Raper
```
Comment (allomov) [inline 2269/2/raper.rb]: having a class Raper is good, but is Battle entity valuable for the given task? should we simply count battles? or is it a distinct entity

btw raper == rapist, rapper == performer

**Comment by allomov** [comment]
the solution does not use all OOP features that were requested.

**Comment by allomov** [comment]
commit names are not perfect, but you are fine to proceed with task #3

**Pull request merged**